CanonicalABI.md

```diff
@@ -0,0 +1,976 @@
+  operation that can be applied to that buffer, for example a `push-buffer` can
+  have items of type `T` pushed into it, and a `pull-buffer` can have items of
+  type `T` removed from it. Buffers have optional capacities (although required
+  in the canonical ABI). The main motivation for buffers is POSIX-like `read`
+  and `write` APIs in WASI. The `read` function, for example, takes a
+  `push-buffer u8` where bytes are pushed into the buffer, and an auxiliary
+  channel (the return value of `read`) indicates how many bytes were read.
+  Buffers are intended to behave similarly to `shared-handle` where they cannot
+  be forged, but unlike `shared-handle` they are only a temporary reference
+  which lives for the duration of a function call (and can only show up as
+  paramters to functions). This enables efficient binding of APIs like `read`
+  and `write` where the canonical ABI is simply passing a pointer/length.
+
+With this grammar of interface types, we can now define what it means to lift
+and lower to all these types.
+
+## Core wasm signatures
+
+First let's take a look at the core wasm signature for any particular interface
+types signature. Note that the core wasm signatures for interface
+types only use the `i32`, `i64`, `f32`, and `f64` wasm types. The `externref`,
+`funcref`, and `v128` types are not used in the core ABI.
+
+```
+# Returns the core wasm signature for a function which takes `params` as
+# arguments (interface types) and returns `results`. The `direction` here is
+# either "import" or "export".
+#
+# Note the usage of a return pointer which is intended to be a temporary
+# workaround for the lack of multi-value today. Languages like C and Rust don't
+# have a great way of idiomatically working with multivalue returns just yet,
+# but this is likely to change in the near future.
+def wasm_signature(direction, params, results):
+    wasm_params = params.flat_map(|t| flatten(direction, t))
+    wasm_results = results.flat_map(|t| flatten(direction, t))
+
+    # If there's more than one result work around the lack of multi-value
+    # returns in C/Rust today. For imports the caller allocates a return pointer
+    # and for exports the callee returns a location in memory where everything
+    # was stored.
+    if len(wasm_results) > 1:
+        if direction == "import"
+          wasm_params.insert(0, i32)
+          wasm_results = []
+        else:
+          wasm_results = [i32]
+
+    return wasm_params, wasm_results
+
+def flatten(direction, interface_type):
+    match interface_type
+        f32 => [f32]
+        f64 => [f64]
+        s8 | s16 | s32 | u8 | u16 | u32 => [i32]
+        s64 | u64 => [i64]
+        char => [i32]
+
+        # List are represented by their pointer/length
+        list $ty => [i32, i32]
+
+        # Records have a specialized representation when they're a
+        # struct-of-bools to represent bitflags. Otherwise it's the flattening
+        # of each member individually
```

Comment:
It seems reasonable to me to not special-case booleans here and go ahead and do packing otherwise. I'm assuming though the packing would only happen for structs-of-enums where the number of bits to represent all the enums is less than 64?